OmniAuth Authorization API

machisuji opened the pull request:
WP [#13781](https://www.openproject.org/work_packages/13781)

machisuji committed:
OmniAuth Authorization API

mfrister commented on app/controllers/concerns/omniauth_login.rb:
```diff
@@ -8,20 +8,34 @@ def omniauth_login
+    user = User.find_or_initialize_by_identity_url identity_url_from_omniauth(auth_hash)
 
-    user = User.find_or_initialize_by_identity_url(identity_url_from_omniauth(auth_hash))
+    dec = OpenProject::OmniAuth::Authorization.authorized? user_with_info(user, auth_hash),
```
What does dec mean? I still don't like abbreviations :wink:

mfrister replied:
Why do you pass the user filled with new data from the authentication provider? The new data might not be valid, so this might prevent changing the user (e.g. if we want to set an admin flag).

mfrister commented on app/controllers/concerns/omniauth_login.rb:
```diff
@@ -8,20 +8,34 @@ def omniauth_login
     end
   end
 
+  def authorization_failed(user, error)
```
Shouldn't this method also be private (like show_error)?

mfrister commented on lib/open_project/omni_auth/authorization.rb:
```diff
@@ -0,0 +1,148 @@
+        rejection = nil
+
+        callbacks.each do |callback|
+          d = callback.authorize user, auth_hash
```
Do we have to use a return value when we pass an object to the callback the user can call a method on?
This might be a bit confusing as one might expect the result being stored on the passed object.

machisuji committed:
do not pass user into authorization callback

machisuji committed:
spell out decision variable's name

machisuji committed:
use decision store so that auth block doesn't necessarily have to return a Decision

machisuji committed:
made rest of support methods private

mfrister commented on config/locales/en.yml:
```diff
@@ -1589,6 +1589,7 @@ en:
     status_user_and_brute_force: "%{user} and %{brute_force}"
     unlock: "Unlock"
     unlock_and_reset_failed_logins: "Unlock and reset failed logins"
+    authorization_rejected: "The user has not been authorized."
```
We could make that (and the German translation) a bit more user-friendly. As this is the message we'll show to the user, we should speak to them directly. E.g. "You're not allowed to sign in." and "Sie dürfen sich nicht einloggen."

machisuji committed:
fixed default rejection; changed message

machisuji committed:
updated comment

machisuji committed:
on_success callback for Authorization

machisuji committed:
corrected/clarified comment

machisuji committed:
pass auth_hash to #on_success. Call it upon bumpy registration too.

mfrister committed:
Omniauth: Add after_login, remove on_success

After a successfull login seems to be a much clearer description of
when the callback will be executed than "on successful authorization".
E.g. it's not clear whether ending up on the registration page also
indicates successful authorization (which it does in the OmniAuth
sense, but not in the OpenProject sense).

The after login callback is called each time a user is logged in, no
matter whether that happens via regular Omniauth login or after
completing the registration process and being automatically logged in
(when automated account activation is enabled).
